File: apps/api/schemas/evals.py
from __future__ import annotations

from pydantic import BaseModel


class EvalRunSummaryItem(BaseModel):
    id: int
    dataset_name: str
    provider: str
    model: str | None
    total_cases: int
    passed_cases: int
    avg_latency_ms: float | None
    created_at: str


class EvalRunCaseItem(BaseModel):
    id: int
    case_id: str
    question: str
    passed: bool
    latency_ms: int
    correctness_score: float | None
    groundedness_score: float | None
    hallucination_score: float | None
    citations: list[str]
    error: str | None


class EvalRunDetail(BaseModel):
    run: EvalRunSummaryItem
    cases: list[EvalRunCaseItem]

File: internal/service/evals.go
package service

import (
	"errors"
)

var ErrEvalRunNotFound = errors.New("Eval run not found")

type EvalRepository interface {
	ListEvalRuns(limit int) ([]EvalRunSummary, error)
	GetEvalRunDetail(evalRunID int) (EvalRunDetailResult, bool, error)
}

type EvalRunSummary struct {
	ID           int
	DatasetName  string
	Provider     string
	Model        *string
	TotalCases   int
	PassedCases  int
	AvgLatencyMS *float64
	CreatedAt    string
}

type EvalRunCase struct {
	ID                 int
	CaseID             string
	Question           string
	Passed             bool
	LatencyMS          int
	CorrectnessScore   *float64
	GroundednessScore  *float64
	HallucinationScore *float64
	Citations          []string
	Error              *string
}

type EvalRunDetailResult struct {
	Run   EvalRunSummary
	Cases []EvalRunCase
}

type EvalService struct {
	repo EvalRepository
}

func NewEvalService(repo EvalRepository) EvalService {
	return EvalService{repo: repo}
}

func (s EvalService) ListEvalRuns(limit int) ([]EvalRunSummary, error) {
	return s.repo.ListEvalRuns(limit)
}

func (s EvalService) GetEvalRunDetail(evalRunID int) (EvalRunDetailResult, error) {
	detail, found, err := s.repo.GetEvalRunDetail(evalRunID)
	if err != nil {
		return EvalRunDetailResult{}, err
	}
	if !found {
		return EvalRunDetailResult{}, ErrEvalRunNotFound
	}
	return detail, nil
}
